tower/src/ready_cache/cache.rs

```diff
@@ -410,20 +431,40 @@ where
     type Output = Result<(K, S, CancelRx), PendingError<K, S::Error>>;
 
     fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
-        let mut fut = self.cancel.as_mut().expect("polled after complete");
-        if let Poll::Ready(r) = Pin::new(&mut fut).poll(cx) {
-            assert!(r.is_ok(), "cancel sender lost");
+        // Before checking whether a service ready, check to see whether
+        // readiness has been canceled.
+        let CancelRx(cancel) = self.cancel.as_mut().expect("polled after complete");
+        if cancel.canceled.load(Ordering::SeqCst) {
             let key = self.key.take().expect("polled after complete");
             return Err(PendingError::Canceled(key)).into();
         }
 
+        // If readiness hasn't been canceled, check to see whether the service
+        // is ready.
         match self
             .ready
             .as_mut()
             .expect("polled after ready")
             .poll_ready(cx)
         {
-            Poll::Pending => Poll::Pending,
+            Poll::Pending => {
+                // Before returning Pending, register interest in cancelation so
+                // that this future is polled again if the state changes.
+                let CancelRx(cancel) = self.cancel.as_mut().expect("polled after complete");
```

Comment:
nit: we already did this above, can't this line be removed?
```suggestion
```

Reply:
```
   --> tower/src/ready_cache/cache.rs:443:15
    |
435 |         let CancelRx(cancel) = self.cancel.as_mut().expect("polled after complete");
    |                                ---- first mutable borrow occurs here
...
443 |         match self
    |               ^^^^ second mutable borrow occurs here
...
453 |                 cancel.waker.register(cx.waker());
    |                 ------ first borrow later used here
```